File: app/src/main/java/com/bubelov/coins/ui/activity/NotificationAreaActivity.java
package com.bubelov.coins.ui.activity;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.SeekBar;

import com.bubelov.coins.Constants;
import com.bubelov.coins.R;
import com.bubelov.coins.model.NotificationArea;
import com.bubelov.coins.repository.area.NotificationAreaRepository;
import com.bubelov.coins.util.OnSeekBarChangeAdapter;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapFragment;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.Circle;
import com.google.android.gms.maps.model.CircleOptions;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Igor Bubelov
 */

public class NotificationAreaActivity extends AbstractActivity implements OnMapReadyCallback {
    public static final String DEFAULT_CAMERA_POSITION_EXTRA = "default_camera_position";

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.bottom_panel)
    View bottomPanel;

    @BindView(R.id.seek_bar_radius)
    SeekBar radiusSeekBar;

    @Inject
    NotificationAreaRepository notificationAreaRepository;

    private GoogleMap map;

    private Circle areaCircle;

    private CameraPosition defaultCameraPosition;

    public static Intent newIntent(Context context, CameraPosition defaultCameraPosition) {
        Intent intent = new Intent(context, NotificationAreaActivity.class);
        intent.putExtra(DEFAULT_CAMERA_POSITION_EXTRA, defaultCameraPosition);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        dependencies().inject(this);
        setContentView(R.layout.activity_select_area);
        ButterKnife.bind(this);

        defaultCameraPosition = getIntent().getParcelableExtra(DEFAULT_CAMERA_POSITION_EXTRA);

        MapFragment mapFragment = (MapFragment) getFragmentManager().findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);

        radiusSeekBar.getProgressDrawable().setColorFilter(getResources().getColor(R.color.accent), PorterDuff.Mode.SRC_IN);
        radiusSeekBar.getThumb().setColorFilter(getResources().getColor(R.color.accent), PorterDuff.Mode.SRC_IN);

        toolbar.setNavigationOnClickListener(v -> {
            saveArea();
            supportFinishAfterTransition();
        });
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        bottomPanel.post(() -> map.setPadding(0, 0, 0, bottomPanel.getHeight()));
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        map = googleMap;
        map.getUiSettings().setMyLocationButtonEnabled(false);
        map.getUiSettings().setZoomControlsEnabled(false);
        map.getUiSettings().setCompassEnabled(false);
        map.setOnMarkerDragListener(new OnMarkerDragListener());

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            map.setMyLocationEnabled(true);
        }

        NotificationArea notificationArea = notificationAreaRepository.getNotificationArea();

        if (notificationArea == null) {
            notificationArea = NotificationArea.builder()
                    .latitude(defaultCameraPosition.target.latitude)
                    .longitude(defaultCameraPosition.target.longitude)
                    .radius(Constants.DEFAULT_NOTIFICATION_AREA_RADIUS_METERS)
                    .build();
        }

        setArea(notificationArea);
    }

    private void setArea(NotificationArea area) {
        BitmapDescriptor markerDescriptor = BitmapDescriptorFactory.fromResource(R.drawable.ic_map_marker_location);

        Marker marker = map.addMarker(new MarkerOptions()
                .position(new LatLng(area.latitude(), area.longitude()))
                .icon(markerDescriptor)
                .anchor(Constants.MAP_MARKER_ANCHOR_U, Constants.MAP_MARKER_ANCHOR_V)
                .draggable(true));

        CircleOptions circleOptions = new CircleOptions()
                .center(marker.getPosition())
                .radius(area.radius())
                .fillColor(getResources().getColor(R.color.notification_area))
                .strokeColor(getResources().getColor(R.color.notification_area_border))
                .strokeWidth(4);

        areaCircle = map.addCircle(circleOptions);

        radiusSeekBar.setMax(500000);
        radiusSeekBar.setProgress((int) areaCircle.getRadius());
        radiusSeekBar.setOnSeekBarChangeListener(new SeekBarChangeListener());

        LatLng areaCenter = new LatLng(area.latitude(), area.longitude());
        map.moveCamera(CameraUpdateFactory.newLatLngZoom(areaCenter, getZoomLevel(areaCircle) - 1));
    }

    private void saveArea() {
        NotificationArea area = NotificationArea.builder()
                .latitude(areaCircle.getCenter().latitude)
                .longitude(areaCircle.getCenter().longitude)
                .radius(areaCircle.getRadius())
                .build();

        notificationAreaRepository.setNotificationArea(area);
    }

    public int getZoomLevel(@NonNull Circle circle) {
        double radius = circle.getRadius();
        double scale = radius / 500;
        return (int) (16 - Math.log(scale) / Math.log(2));
    }

    private class OnMarkerDragListener implements GoogleMap.OnMarkerDragListener {
        @Override
        public void onMarkerDragStart(Marker marker) {
            areaCircle.setFillColor(getResources().getColor(android.R.color.transparent));
        }

        @Override
        public void onMarkerDrag(Marker marker) {
            areaCircle.setCenter(marker.getPosition());
            map.animateCamera(CameraUpdateFactory.newLatLng(marker.getPosition()));
        }

        @Override
        public void onMarkerDragEnd(Marker marker) {
            areaCircle.setFillColor(getResources().getColor(R.color.notification_area));
        }
    }

    private class SeekBarChangeListener extends OnSeekBarChangeAdapter {
        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            areaCircle.setRadius(progress);
        }
    }
}

File: app/src/main/java/com/bubelov/coins/ui/activity/NotificationAreaActivity.kt
/*
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

package com.bubelov.coins.ui.activity

import android.Manifest
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.PorterDuff
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.widget.SeekBar

import com.bubelov.coins.Constants
import com.bubelov.coins.R
import com.bubelov.coins.model.NotificationArea
import com.bubelov.coins.ui.viewmodel.NotificationAreaViewModel
import com.bubelov.coins.util.OnSeekBarChangeAdapter
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapFragment
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.Circle
import com.google.android.gms.maps.model.CircleOptions
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import dagger.android.AndroidInjection

import kotlinx.android.synthetic.main.activity_notification_area.*
import javax.inject.Inject

class NotificationAreaActivity : AppCompatActivity(), OnMapReadyCallback {
    @Inject lateinit var modelFactory: ViewModelProvider.Factory
    private lateinit var model: NotificationAreaViewModel

    private var map: GoogleMap? = null

    private var marker: Marker? = null
    private var areaCircle: Circle? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_notification_area)

        model = ViewModelProviders.of(this, modelFactory)[NotificationAreaViewModel::class.java]

        toolbar.setNavigationOnClickListener {
            saveArea()
            supportFinishAfterTransition()
        }

        val mapFragment = fragmentManager.findFragmentById(R.id.map_fragment) as MapFragment
        mapFragment.getMapAsync(this)

        seek_bar_radius.progressDrawable.setColorFilter(ContextCompat.getColor(this, R.color.accent), PorterDuff.Mode.SRC_IN)
        seek_bar_radius.thumb.setColorFilter(ContextCompat.getColor(this, R.color.accent), PorterDuff.Mode.SRC_IN)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        bottom_panel.post { map!!.setPadding(0, 0, 0, bottom_panel.height) }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        map = googleMap
        map!!.uiSettings.isMyLocationButtonEnabled = false
        map!!.uiSettings.isZoomControlsEnabled = false
        map!!.uiSettings.isCompassEnabled = false
        map!!.setOnMarkerDragListener(OnMarkerDragListener())

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            map!!.isMyLocationEnabled = true
        }

        val defaultCameraPosition = intent.getParcelableExtra<CameraPosition>(DEFAULT_CAMERA_POSITION_EXTRA)

        model.notificationArea.observe(this, Observer { area ->
            if (area == null) {
                model.save(model.getDefaultNotificationArea(defaultCameraPosition))
            } else {
                showArea(area)
            }
        })
    }

    private fun showArea(area: NotificationArea) {
        marker?.remove()
        areaCircle?.remove()

        val markerDescriptor = BitmapDescriptorFactory.fromResource(R.drawable.ic_map_marker_location)

        marker = map!!.addMarker(MarkerOptions()
                .position(LatLng(area.latitude, area.longitude))
                .icon(markerDescriptor)
                .anchor(Constants.MAP_MARKER_ANCHOR_U, Constants.MAP_MARKER_ANCHOR_V)
                .draggable(true))

        val circleOptions = CircleOptions()
                .center(marker?.position)
                .radius(area.radius)
                .fillColor(ContextCompat.getColor(this, R.color.notification_area))
                .strokeColor(ContextCompat.getColor(this, R.color.notification_area_border))
                .strokeWidth(4f)

        areaCircle = map!!.addCircle(circleOptions)

        seek_bar_radius.max = 500000
        seek_bar_radius.progress = areaCircle!!.radius.toInt()
        seek_bar_radius.setOnSeekBarChangeListener(SeekBarChangeListener())

        val areaCenter = LatLng(area.latitude, area.longitude)
        map!!.moveCamera(CameraUpdateFactory.newLatLngZoom(areaCenter, (model.getZoomLevel(areaCircle!!) - 1).toFloat()))
    }

    private fun saveArea() {
        val area = NotificationArea(
                areaCircle!!.center.latitude,
                areaCircle!!.center.longitude,
                areaCircle!!.radius
        )

        model.save(area)
    }

    private inner class OnMarkerDragListener : GoogleMap.OnMarkerDragListener {
        override fun onMarkerDragStart(marker: Marker) {
            areaCircle!!.fillColor = ContextCompat.getColor(this@NotificationAreaActivity, android.R.color.transparent)
        }

        override fun onMarkerDrag(marker: Marker) {
            areaCircle!!.center = marker.position
            map!!.animateCamera(CameraUpdateFactory.newLatLng(marker.position))
        }

        override fun onMarkerDragEnd(marker: Marker) {
            areaCircle!!.fillColor = ContextCompat.getColor(this@NotificationAreaActivity, R.color.notification_area)
            saveArea()
        }
    }

    private inner class SeekBarChangeListener : OnSeekBarChangeAdapter() {
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            areaCircle!!.radius = progress.toDouble()
        }
    }

    companion object {
        const val DEFAULT_CAMERA_POSITION_EXTRA = "default_camera_position"

        fun newIntent(context: Context, defaultCameraPosition: CameraPosition): Intent {
            return Intent(context, NotificationAreaActivity::class.java).apply {
                putExtra(DEFAULT_CAMERA_POSITION_EXTRA, defaultCameraPosition)
            }
        }
    }
}